### PR TITLE
Generalize synopsys dwc2

### LINKED ISSUE #126: Port STM32 L4
**Description**
Port STM32 L4

### LINKED ISSUE #382: Generalize Synopsys DCD/HCD
**Is your feature request related to a problem? Please describe.**
95% of stm32 synopsys and esp32s2 are the same. The differences are mostly register name and address base. Merging them reduce maintaining effort for driver. It is also a good chance to generalize the synopsys driver making adding MCU from other vendor using the same IP easier. 

**Describe the solution you'd like**
merge and rename common driver as `synopsys/dcd_synopsys` 

Ref: linux driver for dwc https://github.com/torvalds/linux/tree/master/drivers/usb/dwc2

MCUs that use synopsys IP
- ST STM32 (F4, F7, H7 etc ..) OTG
- Espressif ESP32 S2/S3
- Giga GDVF103
- Silabs EFM32

### LINKED ISSUE #1141: BCM283x (Raspberry Pi chip) Support
I'm working on adding TinyUSB support for the DWC USB controller from Synopsys in the Broadcom chips on the Raspberry Pi board (BCM2835, BCM2836, BCM2837 and BCM2711). I've haven't managed to get it working so I wanted to open this issue in case others wanted to help get it going.

My working branch is here: https://github.com/tannewt/tinyusb/tree/rpi I'm targeting the compute module 4 on an IO board first because it has a USB micro B plug on it for device mode.

I've modeled the init sequence based on [this](https://github.com/Chadderz121/csud/blob/e13b9355d043a9cdd384b335060f1bc0416df61e/source/hcd/dwc/designware20.c#L583) code but it's for host mode, not device. My first target is the cdc_msc device example. The only other example code I can find for device mode is the Linux kernel and it's quite complicated.

### PR DESCRIPTION
**Pull request by hathach**
**Describe the PR**
Generalize Synopsys DWC2 driver. New driver is at `synopsys/dwc2` For the time being, application can still pick the existing driver e.g st/synopsys/dcd_synopsys.c for stm32. Using this generalized driver is user choice until proven reliably. 

MCUs that use DWC2
- ST STM32 (F4, F7, H7 etc ..) OTG
- Espressif ESP32 S2/S3
- Giga GDVF103
- Silabs EFM32
- Broadcom bcm28/27 on RPi

Ref: linux driver for dwc https://github.com/torvalds/linux/tree/master/drivers/usb/dwc2

- implement and close #382 
- fix #1141 
- fix #126
- #750 should be re-test

### CODE REVIEW

**Comment by hathach** [comment]
somehow the example (cdc_msc) does not work with latest broadcom peripherals since commit 
- https://github.com/adafruit/broadcom-peripherals/commit/0bedbfb7d30d28d4db8790531017ce1b389ebf87

I have tried poking around and making modification where I think it could help. It seems usb phy is initialized but IRQ is not triggered. This proves to be too much knowledge for me digest.

**Comment by tannewt** [comment]
From what I can tell this is a caching issue, not an interrupt one. The first gig of memory is marked as cachable now. If I change it to non-cacheable then the problem goes away. Does the DWC write to or read from any memory directly? It would need to be invalidated or cleaned if so.

```diff
diff --git a/broadcom/mmu.c b/broadcom/mmu.c
index 098c728..3eefaac 100644
--- a/broadcom/mmu.c
+++ b/broadcom/mmu.c
@@ -24,7 +24,7 @@ void setup_mmu_flat_map(void) {
     for (uint64_t i = 1; i < 512 - 8; i++) {
         level_2_0x0_0000_0000_to_0x0_4000_0000[i] = (0x0000000000000000 + (i << 21)) |
                                                     MM_DESCRIPTOR_EXECUTE_NEVER |
-                                                    MM_DESCRIPTOR_MAIR_INDEX(MT_READONLY) |
+                                                    MM_DESCRIPTOR_MAIR_INDEX(MT_NORMAL_NC) |
                                                     MM_DESCRIPTOR_INNER_SHAREABLE |
                                                     MM_DESCRIPTOR_ACCESS_FLAG |
                                                     MM_DESCRIPTOR_BLOCK |
```

**Comment by hathach** [comment]
thanks @tannewt for looking into this, I thought of the cache as well, did a quick check but didn't see any shared memory between cpu and usb. I could miss something, will poke more at that direction. thanks for looking into this +1 +1

**Comment by hathach** [comment]
Thanks to Scott hint, I have narrowed it down to 
- Adding ISB() after `BP_EnableIRQ(USB_IRQn);` fix interrupt not triggered with caching. https://github.com/hathach/tinyusb/pull/1163/commits/b51d038b656688b6150470d23752a306ee150da3#diff-da707f06ba279521a02d7034a0c505b4de440ad706f28c00c4f0cb411ec56373R50
- However while adding Timer1 for LED blinky, I figure out that enable TIMER1 IRQ will auto fix the issue with caching & USB IRQ (above fix is not needed anymore :confused: :confused: ). This explains why circuitpyhon with the same broadcom-peripherals does not has issue with USB when enabling caching.
  
It probably needs more efforts and knowledge to understand the reason behind this, for now I am happy with adding ISB() to dwc2_dcd_int_enable().

File: src/portable/synopsys/dwc2/dwc2_bcm.h [hunk 1]
```diff
@@ -32,28 +32,30 @@
...
+#define dcache_clean(_addr, _size)              data_clean(_addr, _size)
+#define dcache_invalidate(_addr, _size)         data_invalidate(_addr, _size)
+#define dcache_clean_invalidate(_addr, _size)   data_clean_and_invalidate(_addr, _size)
 
 TU_ATTR_ALWAYS_INLINE
 static inline void dwc2_dcd_int_enable(uint8_t rhport)
 {
   (void) rhport;
   BP_EnableIRQ(USB_IRQn);
+  __asm__ volatile("isb");
```
Comment (hathach) [inline src/portable/synopsys/dwc2/dwc2_bcm.h:50]: mentioned in https://github.com/hathach/tinyusb/pull/1163#issuecomment-960473264

File: hw/bsp/raspberrypi4/family.c [hunk 2]
```diff
@@ -50,63 +58,27 @@ void USB_IRQHandler(void)
...
-  // }
-  // while (1) uart_update();
 
   // Turn on USB peripheral.
   vcmailbox_set_power_state(VCMAILBOX_DEVICE_USB_HCD, true);
 
+  // Timer 1/1024 second tick
+  SYSTMR->CS_b.M1 = 1;
+  SYSTMR->C1 = SYSTMR->CLO + 977;
+  BP_EnableIRQ(TIMER_1_IRQn);
```
Comment (hathach) [inline hw/bsp/raspberrypi4/family.c:80]: mentioned in https://github.com/hathach/tinyusb/pull/1163#issuecomment-960473264

**Pull request merged**